Add recipes for Chemfiles

Luthaf opened the pull request:
A C++ library for reading/writing theoretical chemistry trajectories, with Python bindings. See http://chemfiles.github.io/ for more informations. 

I added three recipes: C++ library, Python binding and meta-package. If you prefer having everything in one recipe, I can to it.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/chemfiles```, ```recipes/chemfiles-lib```, ```recipes/chemfiles-python```) and found some lint.

Here's what I've got...


For **recipes/chemfiles**:

 * The top level meta keys are in an unexpected order. Expecting ['package', 'requirements', 'about', 'extra'].
 * The recipe must have some tests.
 * The recipe must have a `build/number` section.

For **recipes/chemfiles-lib**:

 * The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'about', 'extra'].
 * The recipe must have some tests.
 * The recipe must have a `build/number` section.

For **recipes/chemfiles-python**:

 * Failed to even lint the recipe (might be a conda-smithy bug) :cry:

jakirkham commented on recipes/chemfiles-lib/meta.yaml:
```diff
@@ -0,0 +1,36 @@
+  sha256: {{ sha256 }}
+
+requirements:
+  build:
```
Please add `toolchain`. This configures the compiler to use and how to use it.

jakirkham commented on recipes/chemfiles-lib/meta.yaml:
```diff
@@ -0,0 +1,36 @@
+requirements:
+  build:
+    - cmake
+    - libnetcdf
```
Please pin to `4.4.*` based on our [global pinnings]( https://github.com/conda-forge/conda-forge.github.io/blob/e720eaa23843f76bd7dde7e4af46671eead72447/scripts/pin_the_slow_way.py#L41 ).

jakirkham commented on recipes/chemfiles-lib/meta.yaml:
```diff
@@ -0,0 +1,36 @@
+    - cmake
+    - libnetcdf
+  run:
+    - libnetcdf
```
Please also pin to `4.4.*`.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+    - python
+    - enum34  # [py < 34]
+    - numpy
+    - chemfiles-lib ==0.6.0
```
Are these all required at build time? Guessing not. Please drop everything other than `python`. Also please add `setuptools` as that seems to be required for the build.

Luthaf replied:
Yes they are, because the setup.py `import chemfiles` to get the library version.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+      - python setup.py install # [not win]
+  script:  # [win]
+      - "%PYTHON%" setup.py install # [win]
+      - if errorlevel 1 exit 1  # [win]
```
This line can be dropped.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+build:
+  script:  # [not win]
+      - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
+      - python setup.py install # [not win]
```
Please drop the selector and add add the arguments ` --single-version-externally-managed --record=record.txt`.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+  dev_url: https://github.com/chemfiles/chemfiles.py
+
+build:
+  script:  # [not win]
```
Please drop this selector.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+
+build:
+  script:  # [not win]
+      - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
```
Why is this needed? We often set the deployment target with the `toolchain`.

jakirkham commented on recipes/chemfiles-lib/meta.yaml:
```diff
@@ -0,0 +1,36 @@
+    - cmake -DCMAKE_INSTALL_PREFIX=$PREFIX -DENABLE_NETCDF=ON .
+    - cmake --build . --config release --target install
+    - cmake -DBUILD_SHARED_LIBS=ON .
+    - cmake --build . --config release --target install
```
For more complex scripts, we like to break them out into `build.sh` and `bld.bat`.

Luthaf replied:
The point of using cmake here is to have exactly the same commands to build the code in windows and *nix, and so to remove the need for separated build.sh and build.bat files.

jakirkham commented on recipes/chemfiles-lib/meta.yaml:
```diff
@@ -0,0 +1,36 @@
+
+build:
+  script:
+    - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
```
Why is this needed?

Luthaf replied:
I do not remember, but I got an error on Os X when building the recipe a few months ago. I'll check again.

jakirkham commented on recipes/chemfiles-lib/meta.yaml:
```diff
@@ -0,0 +1,36 @@
+
+about:
+  home: http://chemfiles.github.io
+  license: MPL-v2.0
```
Please add `license_file: LICENCE.txt` to include the license file with the package.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+
+about:
+  home: http://chemfiles.github.io
+  license: MPL-v2.0
```
Please add `license_file: LICENCE.txt` to include the license file with the package.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+    - chemfiles-lib ==0.6.0
+  run:
+    - python
+    - enum34  # [py < 34]
```
Please drop the spaces around `<`.

jakirkham commented on recipes/chemfiles-python/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+requirements:
+  build:
+    - python
+    - enum34  # [py < 34]
```
Please drop the spaces around `<`.

jakirkham commented:
Thanks for proposing these, @Luthaf. Glad to see you over here. 😄

Tried to add some addition comments in addition to what the linter included. Once we get these cleaned up a bit more, we can take a closer look.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/chemfiles```, ```recipes/chemfiles-lib```, ```recipes/chemfiles-python```) and found some lint.

Here's what I've got...


For **recipes/chemfiles**:

 * The recipe must have some tests.

For **recipes/chemfiles-lib**:

 * The recipe must have some tests.

For **recipes/chemfiles-python**:

 * The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'test', 'about', 'extra'].
 * The recipe must have a `build/number` section.

Luthaf commented:
I addressed most of the comments; still have to enable tests for chemfiles-lib and check for what is going on with `export MACOSX_DEPLOYMENT_TARGET=""  # [osx]`.

jakirkham commented:
I know the linter is mentioning tests. There are a couple types of tests we like to see.

Running the test suite (mainly for compiled code) so `make check`, `ctest` or similar. Normally that happens in the build.

For Python libraries import tests in the `test` section are nice.

Some checks to make sure files (libraries, headers, configuration files, etc.) were installed where they should be.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/chemfiles```, ```recipes/chemfiles-lib```, ```recipes/chemfiles-python```) and found it was in an excellent condition.

Luthaf commented:
I actually need to `export MACOSX_DEPLOYMENT_TARGET=""`, because without it I get a Bus Error (segfault) in one of my tests. I will try to debug this using the binary built by conda, but I cannot reproduce with a fresh build.

Luthaf commented:
The Circle-CI failure is strange: I got an error about `netcdf.h` file not found, even though `libnetcdf` is part of the `tests.require` entry. Any idea on the origin of this failure?

Luthaf commented:
So after investigations (see the linked issue), the bus error is very weird. I am in favor of just unsetting `MACOSX_DEPLOYMENT_TARGET` in the build, since that make the bug disappear.

@jakirkham, I do not get the origin of the failure on travis/circle-ci, any idea? The Appveyor issue looks strange and related to installing libnetcdf with MSVC.

Luthaf commented:
Appveyor failure seems legitimate, the build log is 

```
conda.resolve.Unsatisfiable: The following specifications were found to be in conflict:
  - cmake
  - libnetcdf 4.4.* -> feature:vc10
  - libnetcdf 4.4.* -> feature:vc14
  - libnetcdf 4.4.* -> feature:vc9
Use "conda info <package>" to see the dependencies for each package.
Command exited with code 1
```

I do not know what I can do here.

Luthaf commented:
Ping @jakirkham? I finally disabled the Windows version, it can be added later.

jakirkham commented:
So I'm in favor of merging.

Though I'd like to save us some time restarting CIs. Could we please put the non-meta recipes (looks like the first 2 commits in a separate PR)? Once done please ping me and I will merge ASAP.

Then we can cut this down the metapackage, which we can merge after the first two are built and deployed.

Sound good?

Luthaf commented:
Sound good, I've separated the non-meta recipes in #1764

jakirkham commented:
Thanks. Have gone ahead and merged that. It might be a little bit as conversion to feedstocks occurs on Travis CI ATM and they are a bit backed up as we speak.

jakirkham commented:
So we ran into some issues with `chemfiles-python`. Have added PR ( https://github.com/conda-forge/chemfiles-python-feedstock/pull/1 ) to fix them. Please take a look. Once that is sorted we should be in good shape to proceed with this.

jakirkham commented:
Toggling for CI.

Luthaf commented:
Should I rebase and only include the meta recipe here?

jakirkham commented:
That sounds like a good idea. Sure let's do that.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/chemfiles```) and found it was in an excellent condition.

jakirkham commented on recipes/chemfiles/meta.yaml:
```diff
@@ -0,0 +1,28 @@
+  version: {{ version }}
+
+build:
+  number: 0
```
Please add `skip: true  # [win]`. That should fix the CI failure. Also feel free to add `[skip appveyor]` in your commit message to avoid having to run that CI as it is not used.

Luthaf commented:
Travis failure looks unrelated (502 gateway error from conda.anaconda.org), could you restart the build?

jakirkham commented:
Yep. Restarted.

jakirkham commented:
Thanks @Luthaf .
